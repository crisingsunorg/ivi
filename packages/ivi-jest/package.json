{
  "private": true,
  "name": "ivi-jest",
  "version": "0.0.1",
  "main": "dist/main/index.js",
  "module": "dist/module/index.js",
  "sideEffects": false,
  "files": [
    "dist",
    "src",
    "!dist/**/*.tsbuildinfo",
    "!src/**/__tests__",
    "README.md",
    "LICENSE"
  ],
  "scripts": {
    "prepublishOnly": "npm run clean && npm run dist",
    "clean": "rm -rf ./dist",
    "dist": "npm run dist:main && npm run dist:module",
    "dist:main": "../../node_modules/.bin/tsc -p ./tsconfig.build.main.json",
    "dist:module": "../../node_modules/.bin/tsc -p ./tsconfig.build.module.json"
  },
  "dependencies": {
    "jsdom": "^14.0.0"
  },
  "peerDependencies": {},
  "devDependencies": {},
  "description": "Jest utils for ivi library",
  "license": "MIT",
  "author": {
    "name": "Boris Kaul",
    "email": "localvoid@gmail.com",
    "url": "https://github.com/localvoid"
  },
  "keywords": [
    "ivi"
  ],
  "homepage": "https://github.com/localvoid/ivi",
  "repository": {
    "type": "git",
    "url": "https://github.com/localvoid/ivi"
  },
  "bugs": {
    "url": "https://github.com/localvoid/ivi/issues"
  },
  "github": "https://github.com/localvoid/ivi"
}
