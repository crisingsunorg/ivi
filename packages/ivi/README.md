ivi is a javascript (TypeScript) library for building web user interfaces.

## Features

- Declarative rendering with "Virtual DOM"
- Components (stateless functions and stateful ES6 classes)
- Connectors for sideways data loading
- Implicit data propagation with contexts
- Extensible synthetic event subsystem
- Synchronous and deterministic syncing algorithm
- Children reconciliation with minimum number of DOM operations
- Fast performance
- Test utilities

## Library Size

Size of the [basic example](https://github.com/localvoid/ivi-examples/tree/master/packages/tutorial/01_introduction)
bundled with [webpack](https://webpack.js.org/) and minified with [uglify](https://github.com/mishoo/UglifyJS2) is just
a **4.4Kb**.

It is possible to get size even lower with [Google Closure Compiler](https://github.com/google/closure-compiler). ivi
library has full support for compilation with Google Closure Compiler in `ADVANCED` mode.

## Performance

ivi is one of the fastest libraries in the [js-framework-benchmark](https://rawgit.com/krausest/js-framework-benchmark/master/webdriver-ts-results/table.html).

It is quite easy to get incredible results in this benchmark just by focusing on optimizations of low-level primitives.
But the primary goal of the ivi architecture is to reduce overhead of abstractions, and that is why it is still one of
the fastest implementations in this benchmark even when benchmark implementation is using stateful components and
connectors.

## Quick Start

The easiest way to get started with ivi is to use [this basic example code on CodePen](https://codepen.io/localvoid/pen/yjqrgj)
or [this one on CodeSandbox](https://codesandbox.io/s/qlypwvz6o6).

The smallest ivi example looks like this:

```js
import { render } from "ivi";
import * as h from "ivi-html";

render(
  h.h1().c("Hello World!"),
  document.getElementById("app"),
);
```

## Documentation

### General

- [Virtual DOM](https://github.com/localvoid/ivi/blob/master/documentation/general/virtual-dom.md)
- [Components](https://github.com/localvoid/ivi/blob/master/documentation/general/components.md)
- [Synthetic Events](https://github.com/localvoid/ivi/blob/master/documentation/general/synthetic-events.md)
- [Connectors](https://github.com/localvoid/ivi/blob/master/documentation/general/connect.md)
- [Context](https://github.com/localvoid/ivi/blob/master/documentation/general/context.md)

### Advanced

- [Scheduler](https://github.com/localvoid/ivi/blob/master/documentation/advanced/scheduler.md)
- [TypeScript Enums](https://github.com/localvoid/ivi/blob/master/documentation/advanced/typescript-enums.md)
- [Security](https://github.com/localvoid/ivi/blob/master/documentation/advanced/security.md)

### Misc

- [Performance Tips](https://github.com/localvoid/ivi/blob/master/documentation/misc/perf-tips.md)
- [Children Reconciliation](https://github.com/localvoid/ivi/blob/master/documentation/misc/children-reconciliation.md)

### Examples and demo applications

#### Boilerplate

- [Basic Boilerplate](https://github.com/localvoid/ivi-boilerplate/)

#### Basic

- [Introduction](https://github.com/localvoid/ivi-examples/tree/master/packages/tutorial/01_introduction/)
- [Stateful Component](https://github.com/localvoid/ivi-examples/tree/master/packages/tutorial/02_stateful_component/)
- [Events](https://github.com/localvoid/ivi-examples/tree/master/packages/tutorial/03_events/)
- [Forms](https://github.com/localvoid/ivi-examples/tree/master/packages/tutorial/04_forms/)
- [Collapsable](https://github.com/localvoid/ivi-examples/tree/master/packages/tutorial/05_collapsable/)

#### Apps

- [TodoMVC](https://github.com/localvoid/ivi-todomvc/)
- [Snake Game](https://github.com/localvoid/ivi-examples/tree/master/packages/apps/snake/)

#### Benchmarks

- [UIBench](https://github.com/localvoid/ivi-examples/tree/master/packages/benchmarks/uibench/)
- [DBMon](https://github.com/localvoid/ivi-examples/tree/master/packages/benchmarks/dbmon/)
- [10k Components](https://github.com/localvoid/ivi-examples/tree/master/packages/benchmarks/10k/)

## License

[MIT](http://opensource.org/licenses/MIT)
