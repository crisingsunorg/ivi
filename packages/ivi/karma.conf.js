const webpack = require("webpack");

module.exports = function (config) {
  config.set({
    browsers: ["ChromeHeadless"],
    frameworks: ["mocha"],
    reporters: ["mocha"],
    preprocessors: { "tests/index.ts": ["webpack", "sourcemap"] },
    files: ["tests/index.ts"],

    colors: true,
    autoWatch: true,

    webpack: {
      module: {
        rules: [
          {
            test: /\.ts$/,
            exclude: /node_modules/,
            use: [
              {
                loader: "ts-loader",
                options: {
                  configFileName: "tsconfig.tests.json",
                },
              },
            ],
          },
        ],
      },
      plugins: [
        new webpack.DefinePlugin({
          "__IVI_DEV__": true,
          "__IVI_BROWSER__": true,
        }),
        new webpack.SourceMapDevToolPlugin({
          test: /\.(ts|js)$/,
        }),
      ],
      resolve: {
        extensions: [".ts", ".js"],
      },
      performance: {
        hints: false
      },
    },

    webpackMiddleware: {
      stats: "errors-only",
      noInfo: true
    },

    mochaReporter: {
      showDiff: true,
    },

    client: {
      mocha: {
        reporter: "html",
        ui: "bdd",
      }
    },

    mime: {
      "text/x-typescript": ["ts"],
    },
  });
};
