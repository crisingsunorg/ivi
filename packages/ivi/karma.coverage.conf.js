const path = require("path");
const webpack = require("webpack");
const merge = require("webpack-merge");
const baseConfig = require("./karma.conf");

module.exports = function (config) {
  baseConfig(config);

  config.set({
    singleRun: true,

    webpack: merge.smart(config.webpack, {
      module: {
        rules: [{
          test: /\.ts$/,
          include: path.resolve("src/"),
          exclude: [
            /node_modules/,
            path.resolve("src/dev_mode/dev_mode.ts"),
            path.resolve("src/dev_mode/stack_trace.ts"),
          ],
          loader: "istanbul-instrumenter-loader",
          enforce: "post",
        }],
      },
    }),

    reporters: ["coverage-istanbul"],

    coverageIstanbulReporter: {
      reports: ["lcov", "text"],
      dir: "./coverage",
      fixWebpackSourcePath: true,
      skipFilesWithNoCoverage: true,
    },
  });

  if (process.env.TRAVIS) {
    config.coverageIstanbulReporter.reports = ["lcov"];
  }
};
