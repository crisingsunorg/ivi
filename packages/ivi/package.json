{
  "name": "ivi",
  "main": "dist/main/index.js",
  "module": "dist/module/index.js",
  "es2016": "dist/es2016/index.js",
  "unpkg": "dist/unpkg/ivi.js",
  "typings": "dist/typings/index.d.ts",
  "sideEffects": false,
  "version": "0.10.0",
  "description": "Javascript (TypeScript) library for building web user interfaces.",
  "license": "MIT",
  "author": {
    "name": "Boris Kaul",
    "email": "localvoid@gmail.com",
    "url": "https://github.com/localvoid"
  },
  "keywords": [
    "ui",
    "virtualdom",
    "vdom",
    "diff",
    "browser",
    "typescript"
  ],
  "homepage": "https://github.com/ivijs/ivi",
  "repository": {
    "type": "git",
    "url": "https://github.com/ivijs/ivi"
  },
  "bugs": {
    "url": "https://github.com/ivijs/ivi/issues"
  },
  "github": "https://github.com/ivijs/ivi",
  "files": [
    "dist",
    "src",
    "README.md",
    "LICENSE"
  ],
  "scripts": {
    "prepublishOnly": "npm run clean && npm run dist",
    "clean": "rm -rf ./dist",
    "dist": "npm run dist:main && npm run dist:module && npm run dist:es2016 && npm run dist:unpkg",
    "dist:main": "../../node_modules/.bin/tsc -p ./tsconfig.build.main.json",
    "dist:module": "../../node_modules/.bin/tsc -p ./tsconfig.build.module.json",
    "dist:es2016": "../../node_modules/.bin/tsc -p ./tsconfig.build.es2016.json",
    "dist:unpkg": "../../node_modules/.bin/webpack --config ./webpack.config.unpkg.js"
  },
  "dependencies": {},
  "peerDependencies": {
    "ivi-core": "^0.10.0",
    "ivi-scheduler": "^0.10.0",
    "ivi-events": "^0.10.0"
  },
  "devDependencies": {
    "ivi-core": "^0.10.0",
    "ivi-scheduler": "^0.10.0",
    "ivi-events": "^0.10.0"
  }
}
