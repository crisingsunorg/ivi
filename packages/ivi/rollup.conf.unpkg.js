import resolve from "rollup-plugin-node-resolve";

export default {
  name: "ivi",
  input: __dirname + "/unpkg/entry.js",
  output: {
    file: __dirname + "/dist/unpkg/index.js",
    format: "umd",
  },
  plugins: [
    resolve(),
  ],
};
