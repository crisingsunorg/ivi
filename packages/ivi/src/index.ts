/**
 * Common types and functions re-exported from ivi-core.
 */
export { KeyCode, KeyLocation, MouseButtons, SyncableValue, PROPERTY, UNSAFE_HTML, EVENT, _ } from "ivi-core";

// Virtual DOM
export { VNodeFlags } from "./vdom/flags";
export { StatelessComponent } from "./vdom/stateless_component";
export { StatefulComponent, Component, isComponentAttached } from "./vdom/stateful_component";
export { ConnectDescriptor } from "./vdom/connect_descriptor";
export {
  VNode,
  getDOMNode, getComponent,
  stopDirtyChecking,
} from "./vdom/vnode";
export { fragment, map, mapRange, mapIterable } from "./vdom/vnode_collections";
export { element } from "./vdom/element";
export { statefulComponent, statelessComponent, withShouldUpdate, context, connect } from "./vdom/vnode_factories";
export { invalidate, dirtyCheck, render } from "./vdom/scheduler";

// Events
export {
  EventHandlerFlags, SyntheticEventFlags, EventFlags, PREVENT_DEFAULT, STOP_PROPAGATION, NativeEventSourceFlags,
} from "./events/flags";
export {
  getEventTarget,
  EVENT_CAPTURE_PASSIVE_OPTIONS, EVENT_CAPTURE_ACTIVE_OPTIONS, EVENT_PASSIVE_OPTIONS, EVENT_ACTIVE_OPTIONS,
} from "./events/utils";
export { SyntheticEvent } from "./events/synthetic_event";
export { EventDispatcher } from "./events/event_dispatcher";
export { EventHandler } from "./events/event_handler";
export { DispatchTarget } from "./events/dispatch_target";
export { accumulateDispatchTargets } from "./events/accumulate_dispatch_targets";
export { dispatchEvent } from "./events/dispatch_event";
export { SyntheticNativeEvent } from "./events/synthetic_native_event";
export {
  NativeEventDispatcher, createNativeEventDispatcher,
  beforeNativeEvent, afterNativeEvent,
  removeBeforeNativeEvent, removeAfterNativeEvent,
} from "./events/native_event_dispatcher";
export {
  EVENT_DISPATCHER_ABORT, EVENT_DISPATCHER_ACTIVATE, EVENT_DISPATCHER_ARIA_REQUEST, EVENT_DISPATCHER_BEFORE_ACTIVATE,
  EVENT_DISPATCHER_BEFORE_COPY, EVENT_DISPATCHER_BEFORE_CUT, EVENT_DISPATCHER_BEFORE_DEACTIVATE,
  EVENT_DISPATCHER_BEFORE_PASTE, EVENT_DISPATCHER_BLUR, EVENT_DISPATCHER_CAN_PLAY, EVENT_DISPATCHER_CAN_PLAYTHROUGH,
  EVENT_DISPATCHER_CHANGE, EVENT_DISPATCHER_CLICK, EVENT_DISPATCHER_CONTEXT_MENU, EVENT_DISPATCHER_COPY,
  EVENT_DISPATCHER_CUE_CHANGE, EVENT_DISPATCHER_CUT, EVENT_DISPATCHER_DOUBLE_CLICK, EVENT_DISPATCHER_DEACTIVATE,
  EVENT_DISPATCHER_DRAG, EVENT_DISPATCHER_DRAG_END, EVENT_DISPATCHER_DRAG_ENTER, EVENT_DISPATCHER_DRAG_LEAVE,
  EVENT_DISPATCHER_DRAG_OVER, EVENT_DISPATCHER_DRAG_START, EVENT_DISPATCHER_DROP, EVENT_DISPATCHER_DURATION_CHANGE,
  EVENT_DISPATCHER_EMPTIED, EVENT_DISPATCHER_ENCRYPTED, EVENT_DISPATCHER_ENDED, EVENT_DISPATCHER_ERROR,
  EVENT_DISPATCHER_FOCUS, EVENT_DISPATCHER_GOT_POINTER_CAPTURE, EVENT_DISPATCHER_INPUT, EVENT_DISPATCHER_INVALID,
  EVENT_DISPATCHER_KEY_DOWN, EVENT_DISPATCHER_KEY_PRESS, EVENT_DISPATCHER_KEY_UP, EVENT_DISPATCHER_LOAD,
  EVENT_DISPATCHER_LOADED_DATA, EVENT_DISPATCHER_LOADED_METADATA, EVENT_DISPATCHER_LOAD_START,
  EVENT_DISPATCHER_LOST_POINTER_CAPTURE, EVENT_DISPATCHER_MOUSE_DOWN, EVENT_DISPATCHER_MOUSE_ENTER,
  EVENT_DISPATCHER_MOUSE_LEAVE, EVENT_DISPATCHER_MOUSE_MOVE, EVENT_DISPATCHER_MOUSE_OUT, EVENT_DISPATCHER_MOUSE_OVER,
  EVENT_DISPATCHER_MOUSE_UP, EVENT_DISPATCHER_PASTE, EVENT_DISPATCHER_PAUSE, EVENT_DISPATCHER_PLAY,
  EVENT_DISPATCHER_PLAYING, EVENT_DISPATCHER_POINTER_CANCEL, EVENT_DISPATCHER_POINTER_DOWN,
  EVENT_DISPATCHER_POINTER_ENTER, EVENT_DISPATCHER_POINTER_LEAVE, EVENT_DISPATCHER_POINTER_MOVE,
  EVENT_DISPATCHER_POINTER_OUT, EVENT_DISPATCHER_POINTER_OVER, EVENT_DISPATCHER_POINTER_UP, EVENT_DISPATCHER_PROGRESS,
  EVENT_DISPATCHER_RATE_CHANGE, EVENT_DISPATCHER_RESET, EVENT_DISPATCHER_SCROLL, EVENT_DISPATCHER_SEEKED,
  EVENT_DISPATCHER_SEEKING, EVENT_DISPATCHER_SELECT, EVENT_DISPATCHER_SELECT_START, EVENT_DISPATCHER_STALLED,
  EVENT_DISPATCHER_SUBMIT, EVENT_DISPATCHER_SUSPEND, EVENT_DISPATCHER_TIME_UPDATE, EVENT_DISPATCHER_TOUCH_CANCEL,
  EVENT_DISPATCHER_TOUCH_END, EVENT_DISPATCHER_TOUCH_MOVE, EVENT_DISPATCHER_TOUCH_START, EVENT_DISPATCHER_UNLOAD,
  EVENT_DISPATCHER_VOLUME_CHANGE, EVENT_DISPATCHER_WAITING, EVENT_DISPATCHER_WHEEL, EVENT_DISPATCHER_ACTIVE_TOUCH_END,
  EVENT_DISPATCHER_ACTIVE_TOUCH_MOVE, EVENT_DISPATCHER_ACTIVE_TOUCH_START, EVENT_DISPATCHER_ACTIVE_WHEEL,

  createNativeEventHandler,

  onAbort, onActivate, onAriaRequest, onBeforeActivate, onBeforeCopy, onBeforeCut, onBeforeDeactivate, onBeforePaste,
  onBlur, onCanPlay, onCanPlaythrough, onChange, onClick, onContextMenu, onCopy, onCueChange, onCut, onDoubleClick,
  onDeactivate, onDrag, onDragEnd, onDragEnter, onDragLeave, onDragOver, onDragStart, onDrop, onDurationChange,
  onEmptied, onEncrypted, onEnded, onError, onFocus, onGotPointerCapture, onInput, onInvalid, onKeyDown, onKeyPress,
  onKeyUp, onLoad, onLoadedData, onLoadedMetadata, onLoadStart, onLostPointerCapture, onMouseDown, onMouseEnter,
  onMouseLeave, onMouseMove, onMouseOut, onMouseOver, onMouseUp, onPaste, onPause, onPlay, onPlaying, onPointerCancel,
  onPointerDown, onPointerEnter, onPointerLeave, onPointerMove, onPointerOut, onPointerOver, onPointerUp, onProgress,
  onRateChange, onReset, onScroll, onSeeked, onSeeking, onSelect, onSelectStart, onStalled, onSubmit, onSuspend,
  onTimeUpdate, onTouchCancel, onTouchEnd, onTouchMove, onTouchStart, onUnload, onVolumeChange, onWaiting, onWheel,
  onActiveTouchEnd, onActiveTouchMove, onActiveTouchStart, onActiveWheel,
} from "./events/native_events";
