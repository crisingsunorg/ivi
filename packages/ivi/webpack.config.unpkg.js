const path = require("path");
const webpack = require("webpack");

module.exports = {
  mode: "production",
  entry: "./src/index.ts",
  output: {
    library: "ivi",
    libraryTarget: "umd",
    filename: "ivi.js",
    path: path.resolve(__dirname, "dist/unpkg"),
  },
  devtool: "source-map",
  optimization: {
    minimize: false,
  },
  externals: {
    "ivi-core": "iviCore",
    "ivi-scheduler": "iviScheduler",
    "ivi-events": "Events",
  },
  module: {
    rules: [
      {
        test: /\.js$/,
        use: ["source-map-loader"],
        enforce: "pre"
      },
      {
        test: /\.ts$/,
        exclude: /node_modules/,
        use: [
          {
            loader: "ts-loader",
            options: {
              configFile: "tsconfig.build.module.json",
            },
          },
        ],
      },
    ],
  },
  plugins: [
    new webpack.DefinePlugin({
      "DEBUG": "true",
      "TARGET": JSON.stringify("browser"),
    }),
  ],
  resolve: {
    extensions: [".ts", ".js"],
  },
};
