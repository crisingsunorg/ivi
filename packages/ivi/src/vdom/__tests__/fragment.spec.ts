import { useComputedValue, useResetDOM, useDOMElement, useIVI, useHTML, useTest } from "ivi-jest";
import { Op } from "ivi";
import { useDOMOpsCounters } from "./jest";

useResetDOM();
const root = useDOMElement();
const domOps = useDOMOpsCounters();
const ivi = useIVI();
const h = useHTML();
const t = useTest();
const Component = useComputedValue(() => ivi.component<Op>(() => (ops) => ops));
const r = (op: Op) => t.render(op, root());

describe("fragment", () => {
  describe("mount", () => {
    test("empty", () => {
      r([]);
      expect(root()).toMatchSnapshot();
      expect(domOps()).toMatchSnapshot();
    });

    test("single child", () => {
      r([h.h1()]);
      expect(root()).toMatchSnapshot();
      expect(domOps()).toMatchSnapshot();
    });

    test("two children", () => {
      r([h.h1(), h.h2()]);
      expect(root()).toMatchSnapshot();
      expect(domOps()).toMatchSnapshot();
    });

    test("null between two children", () => {
      r([h.h1(), null, h.h2()]);
      expect(root()).toMatchSnapshot();
      expect(domOps()).toMatchSnapshot();
    });

    describe("components", () => {
      test("first null", () => {
        r([Component(null), Component(h.h1()), Component(h.h2())]);
        expect(root()).toMatchSnapshot();
        expect(domOps()).toMatchSnapshot();
      });

      test("last null", () => {
        r([Component(h.h1()), Component(h.h2()), Component(null)]);
        expect(root()).toMatchSnapshot();
        expect(domOps()).toMatchSnapshot();
      });

      test("null between two children", () => {
        r([Component(h.h1()), Component(null), Component(h.h2())]);
        expect(root()).toMatchSnapshot();
        expect(domOps()).toMatchSnapshot();
      });
    });
  });

  describe("update", () => {
    describe("general", () => {
      test("null to two children", () => {
        r(null);
        r([h.h1(), h.h2()]);
        expect(root()).toMatchSnapshot();
        expect(domOps()).toMatchSnapshot();
      });

      test("null to fragment with a hole", () => {
        r(null);
        r([null]);
        expect(root()).toMatchSnapshot();
        expect(domOps()).toMatchSnapshot();
      });

      test("one child to two children", () => {
        r([h.h1()]);
        r([h.h1(), h.h2()]);
        expect(root()).toMatchSnapshot();
        expect(domOps()).toMatchSnapshot();
      });

      test("two children to one child", () => {
        r([h.h1(), h.h2()]);
        r([h.h1()]);
        expect(root()).toMatchSnapshot();
        expect(domOps()).toMatchSnapshot();
      });

      test("two children to zero children", () => {
        r([h.h1(), h.h2()]);
        r([]);
        expect(root()).toMatchSnapshot();
        expect(domOps()).toMatchSnapshot();
      });

      test("hole 1", () => {
        r([null, null]);
        r([h.h1(), null]);
        expect(root()).toMatchSnapshot();
        expect(domOps()).toMatchSnapshot();
      });

      test("hole 2", () => {
        r([null, null]);
        r([null, h.h2()]);
        expect(root()).toMatchSnapshot();
        expect(domOps()).toMatchSnapshot();
      });

      test("hole 3", () => {
        r([null, null]);
        r([h.h1(), h.h2()]);
        expect(root()).toMatchSnapshot();
        expect(domOps()).toMatchSnapshot();
      });

      test("hole 4", () => {
        r([h.h1(), null]);
        r([null, null]);
        expect(root()).toMatchSnapshot();
        expect(domOps()).toMatchSnapshot();
      });

      test("hole 5", () => {
        r([null, h.h2()]);
        r([null, null]);
        expect(root()).toMatchSnapshot();
        expect(domOps()).toMatchSnapshot();
      });

      test("hole 6", () => {
        r([h.h1(), h.h2()]);
        r([null, null]);
        expect(root()).toMatchSnapshot();
        expect(domOps()).toMatchSnapshot();
      });

      test("hole 7", () => {
        r([h.h1(), null]);
        r([null, h.h2()]);
        expect(root()).toMatchSnapshot();
        expect(domOps()).toMatchSnapshot();
      });

      test("hole 8", () => {
        r([null, h.h2()]);
        r([h.h1(), null]);
        expect(root()).toMatchSnapshot();
        expect(domOps()).toMatchSnapshot();
      });
    });

    describe("move", () => {
      test("1", () => {
        r(ivi.TrackByKey([ivi.key(0, [1, 2]), ivi.key(1, [3, 4])]));
        r(ivi.TrackByKey([ivi.key(1, [3, 4]), ivi.key(0, [1, 2])]));
        expect(root()).toMatchSnapshot();
        expect(domOps()).toMatchSnapshot();
      });

      test("2", () => {
        const n0 = ivi.key(0, [1, 2]);
        const n1 = ivi.key(1, [3, 4]);
        r(ivi.TrackByKey([n0, n1]));
        r(ivi.TrackByKey([n1, n0]));
        expect(root()).toMatchSnapshot();
        expect(domOps()).toMatchSnapshot();
      });
    });
  });
});
