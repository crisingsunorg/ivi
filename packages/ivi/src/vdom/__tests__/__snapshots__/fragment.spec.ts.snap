// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fragment mount components first null 1`] = `
<div>
  <h1 />
  <h2 />
</div>
`;

exports[`fragment mount components first null 2`] = `
Object {
  "appendChild": 0,
  "createElement": 2,
  "createElementNS": 0,
  "createTextNode": 0,
  "innerHTML": 0,
  "insertBefore": 2,
  "nodeValue": 0,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`fragment mount components last null 1`] = `
<div>
  <h1 />
  <h2 />
</div>
`;

exports[`fragment mount components last null 2`] = `
Object {
  "appendChild": 0,
  "createElement": 2,
  "createElementNS": 0,
  "createTextNode": 0,
  "innerHTML": 0,
  "insertBefore": 2,
  "nodeValue": 0,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`fragment mount components null between two children 1`] = `
<div>
  <h1 />
  <h2 />
</div>
`;

exports[`fragment mount components null between two children 2`] = `
Object {
  "appendChild": 0,
  "createElement": 2,
  "createElementNS": 0,
  "createTextNode": 0,
  "innerHTML": 0,
  "insertBefore": 2,
  "nodeValue": 0,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`fragment mount empty 1`] = `<div />`;

exports[`fragment mount empty 2`] = `
Object {
  "appendChild": 0,
  "createElement": 0,
  "createElementNS": 0,
  "createTextNode": 0,
  "innerHTML": 0,
  "insertBefore": 0,
  "nodeValue": 0,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`fragment mount null between two children 1`] = `
<div>
  <h1 />
  <h2 />
</div>
`;

exports[`fragment mount null between two children 2`] = `
Object {
  "appendChild": 0,
  "createElement": 2,
  "createElementNS": 0,
  "createTextNode": 0,
  "innerHTML": 0,
  "insertBefore": 2,
  "nodeValue": 0,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`fragment mount single child 1`] = `
<div>
  <h1 />
</div>
`;

exports[`fragment mount single child 2`] = `
Object {
  "appendChild": 0,
  "createElement": 1,
  "createElementNS": 0,
  "createTextNode": 0,
  "innerHTML": 0,
  "insertBefore": 1,
  "nodeValue": 0,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`fragment mount two children 1`] = `
<div>
  <h1 />
  <h2 />
</div>
`;

exports[`fragment mount two children 2`] = `
Object {
  "appendChild": 0,
  "createElement": 2,
  "createElementNS": 0,
  "createTextNode": 0,
  "innerHTML": 0,
  "insertBefore": 2,
  "nodeValue": 0,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`fragment update general hole 1 1`] = `
<div>
  <h1 />
</div>
`;

exports[`fragment update general hole 1 2`] = `
Object {
  "appendChild": 0,
  "createElement": 1,
  "createElementNS": 0,
  "createTextNode": 0,
  "innerHTML": 0,
  "insertBefore": 1,
  "nodeValue": 0,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`fragment update general hole 2 1`] = `
<div>
  <h2 />
</div>
`;

exports[`fragment update general hole 2 2`] = `
Object {
  "appendChild": 0,
  "createElement": 1,
  "createElementNS": 0,
  "createTextNode": 0,
  "innerHTML": 0,
  "insertBefore": 1,
  "nodeValue": 0,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`fragment update general hole 3 1`] = `
<div>
  <h1 />
  <h2 />
</div>
`;

exports[`fragment update general hole 3 2`] = `
Object {
  "appendChild": 0,
  "createElement": 2,
  "createElementNS": 0,
  "createTextNode": 0,
  "innerHTML": 0,
  "insertBefore": 2,
  "nodeValue": 0,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`fragment update general hole 4 1`] = `<div />`;

exports[`fragment update general hole 4 2`] = `
Object {
  "appendChild": 0,
  "createElement": 1,
  "createElementNS": 0,
  "createTextNode": 0,
  "innerHTML": 0,
  "insertBefore": 1,
  "nodeValue": 0,
  "removeChild": 1,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`fragment update general hole 5 1`] = `<div />`;

exports[`fragment update general hole 5 2`] = `
Object {
  "appendChild": 0,
  "createElement": 1,
  "createElementNS": 0,
  "createTextNode": 0,
  "innerHTML": 0,
  "insertBefore": 1,
  "nodeValue": 0,
  "removeChild": 1,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`fragment update general hole 6 1`] = `<div />`;

exports[`fragment update general hole 6 2`] = `
Object {
  "appendChild": 0,
  "createElement": 2,
  "createElementNS": 0,
  "createTextNode": 0,
  "innerHTML": 0,
  "insertBefore": 2,
  "nodeValue": 0,
  "removeChild": 2,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`fragment update general hole 7 1`] = `
<div>
  <h2 />
</div>
`;

exports[`fragment update general hole 7 2`] = `
Object {
  "appendChild": 0,
  "createElement": 2,
  "createElementNS": 0,
  "createTextNode": 0,
  "innerHTML": 0,
  "insertBefore": 2,
  "nodeValue": 0,
  "removeChild": 1,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`fragment update general hole 8 1`] = `
<div>
  <h1 />
</div>
`;

exports[`fragment update general hole 8 2`] = `
Object {
  "appendChild": 0,
  "createElement": 2,
  "createElementNS": 0,
  "createTextNode": 0,
  "innerHTML": 0,
  "insertBefore": 2,
  "nodeValue": 0,
  "removeChild": 1,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`fragment update general null to fragment with a hole 1`] = `<div />`;

exports[`fragment update general null to fragment with a hole 2`] = `
Object {
  "appendChild": 0,
  "createElement": 0,
  "createElementNS": 0,
  "createTextNode": 0,
  "innerHTML": 0,
  "insertBefore": 0,
  "nodeValue": 0,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`fragment update general null to two children 1`] = `
<div>
  <h1 />
  <h2 />
</div>
`;

exports[`fragment update general null to two children 2`] = `
Object {
  "appendChild": 0,
  "createElement": 2,
  "createElementNS": 0,
  "createTextNode": 0,
  "innerHTML": 0,
  "insertBefore": 2,
  "nodeValue": 0,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`fragment update general one child to two children 1`] = `
<div>
  <h1 />
  <h2 />
</div>
`;

exports[`fragment update general one child to two children 2`] = `
Object {
  "appendChild": 0,
  "createElement": 2,
  "createElementNS": 0,
  "createTextNode": 0,
  "innerHTML": 0,
  "insertBefore": 2,
  "nodeValue": 0,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`fragment update general two children to one child 1`] = `
<div>
  <h1 />
</div>
`;

exports[`fragment update general two children to one child 2`] = `
Object {
  "appendChild": 0,
  "createElement": 2,
  "createElementNS": 0,
  "createTextNode": 0,
  "innerHTML": 0,
  "insertBefore": 2,
  "nodeValue": 0,
  "removeChild": 1,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`fragment update general two children to zero children 1`] = `<div />`;

exports[`fragment update general two children to zero children 2`] = `
Object {
  "appendChild": 0,
  "createElement": 2,
  "createElementNS": 0,
  "createTextNode": 0,
  "innerHTML": 0,
  "insertBefore": 2,
  "nodeValue": 0,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 1,
}
`;

exports[`fragment update move 1 1`] = `
<div>
  3
  4
  1
  2
</div>
`;

exports[`fragment update move 1 2`] = `
Object {
  "appendChild": 0,
  "createElement": 0,
  "createElementNS": 0,
  "createTextNode": 4,
  "innerHTML": 0,
  "insertBefore": 6,
  "nodeValue": 0,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`fragment update move 2 1`] = `
<div>
  3
  4
  1
  2
</div>
`;

exports[`fragment update move 2 2`] = `
Object {
  "appendChild": 0,
  "createElement": 0,
  "createElementNS": 0,
  "createTextNode": 4,
  "innerHTML": 0,
  "insertBefore": 6,
  "nodeValue": 0,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 0,
}
`;
